baybe/utils/basic.py

```diff
@@ -181,3 +182,37 @@ def find_subclass(base: type, name_or_abbr: str, /):
             f"The class name or abbreviation '{name_or_abbr}' does not refer to any "
             f"of the subclasses of '{base.__name__}'."
         )
+
+
+def register_hook(target: Callable, hook: Callable) -> Callable:
+    """Register a callable for the given target method.
+
+    Args:
+        target: The method to be changed.
+        hook: The callable to be registered in the given method.
+
+    Returns:
+        A wrapped callable that replaces the original target method.
+
+    Raises:
+        TypeError: If the signature of the callable does not match the signature of the
+            target method.
+    """
+    for p1, p2 in zip(
```

Comment:
hmm I think we said that the hook must accept i) the same sig as the original method --> implemented but also ii) other attributes that are available to the method

without ii) the user hooks wont be able to calculate/infer everything that the original method also does

Reply:
So do you suggest to return all attributes that are available for the target method (such as: `surrogate_model` and `acquisition_function` for a `BayesianRecommender`) in form of a list or to make it possible for the hook function to have more attributes in it's signature?

Reply:
so we have two things i) signature arguments of the method and ii) object attributes of the class the method belongs to

There are several complications arising from that that are not reflected in this PR. The things that would need to chagne then are:

1) You also need to get ii) and add them to the **kwargs you pass. I think in our meeting I wrote something like `return hook(*args, **kwargs, **attributes)`. Not sure if thats the best way but they also need to be passed to the hook somehow

2) Because of that you cannot make a simple signature matching between the hook and the method. Perhaps a check whether the kwargs of the hook cover all kwargs+attributes of the method+class, but the order imo doesnt matter. The matching for `args` would still remain the same. OR we simply say the hook only has to follow the `args` singature of the method, the rest could be covered by it having `**kwargs`. Eg a hook can also be coded like `def myhook(arg1, arg2, arg3, **kwargs): print(kwargs['one_of_the_kwargs_of_the_method_or_an_attribute_of_the_obejct'])` (arg1-3 would need to match the signature args of the overwritten method/function)

3) 1 and 2 are only relevant for the case the function is a method. In case a pure function is overwritten then they dont apply. I think overwreiting methods is muh more itneresting though than isolated functions. If you want to support both, the utility needs to make the corresponding additional actions / checks based on whether the callable passed is a method or just a function.

Reply:
Okay I will try to find a way to make the attributes also accessable.
But according to 3: do you really want to make this feature available for pure functions as well? Normally all attributes a function has, would be given by the user, right? Or do I understand it wrong?

Reply:
Sorry, perhaps I'm not following completely, but didn't we agree that hook simply also gets `self` (i.e. the object itself)!? That would be the easiest / most natural way, and no complicated matching necessary!

Reply:
In fact, @RimRihana's example already registers the hook with a method, and everything seems to work as intended. When you mess up the signature, e.g. change the first part to `self: str`, it correctly raises an error

Reply:
Yes, that is true. I think this is the easiest way to access  the private attributes. 
Example: If you try to print `self` inside the callable function, you will get the object printed with it's private attributes like this: 
![image](https://github.com/emdgroup/baybe/assets/118910837/4a1cae8f-2b36-41c3-a4f9-bd0d355035c4)

Reply:
indeed with self + args/kwargs both situations I) and ii) are served and its laso easier to code a hook with exactly the right signature (namely self + args/kwargs eg taken from the doc)